### PR TITLE
Add Runtime metaprogramming, GroovyObject and GroovyInterceptable section to Groovy documentation

### PR DESCRIPTION
**Pull request by stavytskyi**

### CODE REVIEW

**Comment by stavytskyi** [comment]
Dear colleagues, please review my results. It's my first contribution so If you have some remarks or suggestion please tell me.

**Comment by blackdrag** [comment]
you can please rebase your commit to current master? Then we can merge this more easily

**Comment by stavytskyi** [comment]
Sorry, no problem.

On Mon, Mar 30, 2015 at 12:00 PM, Jochen Theodorou <notifications@github.com

**Comment by stavytskyi** [comment]
I have tested tests which I have added during this commits and they passed. @blackdrag  Could you help with this problem?

**Comment by stavytskyi** [comment]
I think this problem related to Indy build on JDK7. I haven't this problem on JDK8. If I wrote wrong test, please, inform me.

File: src/spec/test/metaprogramming/InterceptionThroughMetaClassTest.groovy [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+import groovy.xml.Entity
+
+// tag::meta_class_interception[]
+class InterceptionThroughMetaClassTest extends GroovyTestCase {
+    void testPOJOMetaClassInterception() {
+        String invoking = 'ha'
+        invoking.metaClass.invokeMethod = {String name, Object args ->
+            'invoked'
+        }
+        assertEquals invoking.length(), invoking.someMethod()
```
Comment (PascalSchumacher) [inline src/spec/test/metaprogramming/InterceptionThroughMetaClassTest.groovy]: I'm no expert on metaprogramming, but I'm surprised that this test seems to work on Jdk8. I would have expected it to fail like this http://ci.groovy-lang.org/viewLog.html?buildId=16276&buildTypeId=Groovy_Jdk7Build every time. 

I suggest using `assert invoking.someMethod() == 'invoked'` instead. Imho this makes the intent a bit clearer.

File: src/spec/test/metaprogramming/InterceptionThroughMetaClassTest.groovy [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+    }
+
+    void testPOGOMetaClassInterception() {
+        Entity entity = new Entity('Hello')
+        entity.metaClass.invokeMethod = {String name, Object args ->
+            'invoked'
+        }
+        assertEquals entity.build(new Object()), entity.someMethod()
```
Comment (PascalSchumacher) [inline src/spec/test/metaprogramming/InterceptionThroughMetaClassTest.groovy]: I would also change this to `assert invoking.someMethod() == 'invoked'` instead. Imho this makes the intent a bit clearer. What do you think?

**Comment by stavytskyi** [comment]
@PascalSchumacher I would like to show that my code changes default behavior of exist POJO(String) and POGO(Entity) classes. In both cases all method(both exist and non exist) invocations should return `invoked` so we can change `assertEquals invoking.length(), invoking.someMethod()` to:
```groovy
        assert invoking.length() == 'invoked'
        assert invoking.someMethod() == 'invoked'
```
and
` assertEquals entity.build(new Object()), entity.someMethod()` to:
```groovy
        assert entity.build(new Object()) == 'invoked'
        assert entity.someMethod() == 'invoked'
```
for easy reading. 
I have checked this changes (on top) and they failed too only on JDK7 with Indy option.
Something wrong in `Groovy MOP` in *Indy* build only for JDK7. So in my opinion either it is a bug or undefined behavior.

**Comment by PascalSchumacher** [comment]
@stavytskyi I see. You are right. Sorry about not getting your intention.

@blackdrag I guess you have to help us out.

**Comment by blackdrag** [comment]
See https://issues.apache.org/jira/browse/GROOVY-7391 I am trying to push a fix tonight

**Comment by PascalSchumacher** [comment]
Merged: https://github.com/apache/incubator-groovy/commit/2d42a54df73b6ca0410d10b1ac5fdb60fe23edfd https://github.com/apache/incubator-groovy/commit/16b42217c840bf1be98f203e02992981534fa0da

Thanks!

**Pull request closed without merging**